### PR TITLE
Add CodeQL code scanning

### PR DESCRIPTION
**Pull request by tristan957**
CodeQL is a static analysis tool from GitHub that can find common
mistakes in code that in some cases can lead to security issues.

Signed-off-by: Tristan Partin <tpartin@micron.com>

### CODE REVIEW

**Comment by beckerg** [comment]
What's the procedure for nullifying a false-positive?

**Review by beckerg — APPROVED** [review]

**Comment by tristan957** [comment]
I am just submitting this to see what it does. Not sure what the answer to your question is at the moment. I don't really intend to merge this without some discussion.

**Comment by beckerg** [comment]
Right, will be curious to see what it turns up.  Some such tools will report a thousand issues on first run, and someone has to wade through them to find the three that are valid.  So maybe this tool is much better in that regard, won't know until we try it...

**Comment by tristan957** [comment]
Here is the first draft.

https://github.com/hse-project/hse/security/code-scanning?query=pr%3A354+is%3Aopen

There are some good ones in there, like: https://github.com/hse-project/hse/security/code-scanning/27

Some of these are also things that clang-tidy can catch, like these format string ones. I would like to get that one integrated at some point. Task for another day.

I am not really too well versed in what checks are available vs what checks are currently enabled. Most issues it found arose in tests and tools, which have at times been less than stellar code.

The only stupid check I can see is `%p` has to be a `void *`, which honestly I have never understood, but it is an easy fix.

**Comment by alexttx** [comment]
> Bullseye (was it bullseye?)

You're probably thinking of Coverity.

**Comment by tristan957** [comment]
> Some of these are also things that clang-tidy can catch, like these format string ones. I would like to get that one integrated at some point. Task for another day.

Another day it was...

**Comment by alexttx** [comment]
Seems like we need to annotate more functions with `HSE_PRINTF(...)`.   That would catch all the format string errors at compile time.

**Comment by tristan957** [comment]
Yea that function attribute is really quite amazing. I have been using it for some of the new rest code.

File: lib/cn/vblock_reader.c [hunk 1]
```diff
@@ -98,10 +98,10 @@ vbr_desc_update(
 void
 vbr_readahead(
     struct vblock_desc *     vbd,
-    u32                      voff,
-    u32                      vlen,
+    off_t                    voff,
+    size_t                   vlen,
     u32                      ra_flags,
-    u32                      ra_len,
+    size_t                   ra_len,
```
Comment (beckerg) [inline lib/cn/vblock_reader.c:104]: voff is now signed, i hope there isn't any issue with that...

File: tools/kmt/kmt.c [hunk 2]
```diff
@@ -247,11 +247,11 @@ bool           latency = false;
...
 bool           keybinary = false;
 char          *fieldname_fmt = "field%u";
-uint           fieldnamew_min; /* minimum fieldname width */
-uint           fieldnamew_max; /* maximum fieldname width */
+size_t         fieldnamew_min; /* minimum fieldname width */
+size_t         fieldnamew_max; /* maximum fieldname width */
 uint           fieldcount_max = 2048;
 uint           fieldcount = 1;
-uint           fieldlength = 0;
+size_t         fieldlength = 0;
```
Comment (beckerg) [inline tools/kmt/kmt.c:254]: I think these should all be signed int to avoid breaking tests of them against the int returned by snprintf()...

**Review by beckerg — APPROVED** [review]

File: lib/cn/vblock_reader.c (hunk 1 above)
Comment (tristan957) [reply lib/cn/vblock_reader.c:104]: reading the code, shit would hit the fan if voff was signed

Comment (tristan957) [reply lib/cn/vblock_reader.c:104]: Let me just revert that

**Review by alexttx — APPROVED** [review]

**Pull request merged**